### PR TITLE
XRT-186 data retention new flow

### PR DESCRIPTION
**Pull request by chienwei-lan**
Data Retention new flow.

1. ulp isolate sequence (0x3 => 0x1 => 0x0), undo (0x0 => 0x1 => 0x3)

2. word-addressing XSDB

3. move icap_save_calib before destroy ULP level sub-devices

### CODE REVIEW

**Comment by houlz0507** [comment]
@chienwei-lan I did not see issue. Could you try u50 and see if hello world works?

**Comment by xdavidz** [comment]
> Data Retention new flow.
> 
> 1. ulp isolate sequence (0x3 => 0x1 => 0x0), undo (0x0 => 0x1 => 0x3)
> 2. word-addressing XSDB
> 3. move icap_save_calib before destroy ULP level sub-devices

I don't have background to review this webrev.

**Comment by chienwei-lan** [comment]
@houlz0507  

[ 9228.520425] xclmgmt 0000:b3:00.0:  ffff880444453098 xclmgmt_probe: Driver: 2.6.0,**e808fa68ae64bee732f7a3c24bae65549289641a**
[ 9228.520431] xclmgmt 0000:b3:00.0:  ffff880444453098 xclmgmt_probe: probe(pdev = 0xffff880444453000, pci_id = 0xffffffffc07abe60)
[ 9228.522556] xclmgmt 0000:b3:00.0:  ffff880444453098 xclmgmt_probe: minimum initialization done
root@xsjlizhih50:/tmp/out/usr/src/xrt-2.6.0/driver/xocl/userpf# ^C
root@xsjlizhih50:/tmp/out/usr/src/xrt-2.6.0/driver/xocl/userpf# xbutil validate -q
INFO: Found 1 cards

INFO: Validating card[0]: xilinx_u50_gen3x16_xdma_201920_3
INFO: == Starting AUX power connector check:
AUX power connector not available. Skipping validation
INFO: == AUX power connector check SKIPPED
INFO: == Starting PCIE link check:
INFO: == PCIE link check PASSED
INFO: == Starting SC firmware version check:
INFO: == SC firmware version check PASSED
INFO: == Starting verify kernel test:
INFO: == verify kernel test PASSED
INFO: Card[0] validated successfully.

INFO: All cards validated successfully.

**Comment by xdavidz** [comment]
> @houlz0507
> 
> [ 9228.520425] xclmgmt 0000:b3:00.0: ffff880444453098 xclmgmt_probe: Driver: 2.6.0,**[e808fa6](https://github.com/Xilinx/XRT/commit/e808fa68ae64bee732f7a3c24bae65549289641a)**
> [ 9228.520431] xclmgmt 0000:b3:00.0: ffff880444453098 xclmgmt_probe: probe(pdev = 0xffff880444453000, pci_id = 0xffffffffc07abe60)
> [ 9228.522556] xclmgmt 0000:b3:00.0: ffff880444453098 xclmgmt_probe: minimum initialization done
> root@xsjlizhih50:/tmp/out/usr/src/xrt-2.6.0/driver/xocl/userpf# ^C
> root@xsjlizhih50:/tmp/out/usr/src/xrt-2.6.0/driver/xocl/userpf# xbutil validate -q
> INFO: Found 1 cards
> 
> INFO: Validating card[0]: xilinx_u50_gen3x16_xdma_201920_3
> INFO: == Starting AUX power connector check:
> AUX power connector not available. Skipping validation
> INFO: == AUX power connector check SKIPPED
> INFO: == Starting PCIE link check:
> INFO: == PCIE link check PASSED
> INFO: == Starting SC firmware version check:
> INFO: == SC firmware version check PASSED
> INFO: == Starting verify kernel test:
> INFO: == verify kernel test PASSED
> INFO: Card[0] validated successfully.
> 
> INFO: All cards validated successfully.

could you please run xbutil program -p bandwidth.xclbin too?

**Comment by chienwei-lan** [comment]
/tmp/out/usr/src/xrt-2.6.0/driver/xocl/userpf# xbutil validate
INFO: Found 1 cards

INFO: Validating card[0]: xilinx_u50_gen3x16_xdma_201920_3
INFO: == Starting AUX power connector check:
AUX power connector not available. Skipping validation
INFO: == AUX power connector check SKIPPED
INFO: == Starting PCIE link check:
INFO: == PCIE link check PASSED
INFO: == Starting SC firmware version check:
INFO: == SC firmware version check PASSED
INFO: == Starting verify kernel test:
INFO: == verify kernel test PASSED
INFO: == Starting DMA test:
Host -> PCIe -> FPGA write bandwidth = 8015.03 MB/s
Host <- PCIe <- FPGA read bandwidth = 7221.44 MB/s
INFO: == DMA test PASSED
INFO: == Starting device memory bandwidth test:

...........
Maximum throughput: 52428 MB/s
INFO: == device memory bandwidth test PASSED
INFO: == Starting PCIE peer-to-peer test:
P2P BAR is not enabled. Skipping validation
INFO: == PCIE peer-to-peer test SKIPPED
INFO: == Starting memory-to-memory DMA test:
M2M is not available. Skipping validation
INFO: == memory-to-memory DMA test SKIPPED
INFO: Card[0] validated successfully.

INFO: All cards validated successfully.

File: src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c [hunk 1]
```diff
@@ -25,9 +25,8 @@
 #define	REG_CALIB_OFFSET		0x00000008
 #define	REG_XSDB_RAM_BASE		0x00004000
 
-#define	XSDB_RAM_MAX_IN_WORDS		2901
+#define	XSDB_RAM_MAX_IN_WORDS		2902
```
Comment (maxzhen) [inline src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: we still can't get this number from HW?

Comment (chienwei-lan) [reply src/runtime_src/core/pcie/driver/linux/xocl/subdev/ddr_srsr.c]: done, get the size from XSDB offset +4 & +8

**Pull request merged**